File: day18/day18.py
#!/usr/bin/env python3

# tokens
NUM = 0
ADD = 1
MUL = 2
LPAR = 3
RPAR = 4
EOF = 5

# expression = factor {"+" factor | "*" factor}*
# factor = number | "(" expression ")"
class Grammar1(object):
    def __init__(self):
        self.pos = 0
        self.cur = (EOF, None)
        self.txt = ""

    def eval(self, txt):
        self.pos = 0
        self.txt = txt
        self.scan_next()
        return self.expression()

    def scan_next(self):
        while self.pos < len(self.txt) and self.txt[self.pos] in " \n\t\r":
            self.pos +=1
        if self.pos == len(self.txt):
            self.cur = (EOF, None)
        elif self.txt[self.pos] in "0123456789":
            start = self.pos
            while self.pos < len(self.txt) and self.txt[self.pos] in "0123456789":
                self.pos += 1
            self.cur = (NUM, self.txt[start:self.pos])
        elif self.txt[self.pos] == "+":
            self.cur = (ADD, None)
            self.pos += 1
        elif self.txt[self.pos] == "*":
            self.cur = (MUL, None)
            self.pos += 1
        elif self.txt[self.pos] == "(":
            self.cur = (LPAR, None)
            self.pos += 1
        elif self.txt[self.pos] == ")":
            self.cur = (RPAR, None)
            self.pos += 1
        else:
            print("Error: {} at {}".format(self.txt, self.pos))
            exit(1)

    # factor = number | "(" expression ")"
    def factor(self):
        if self.cur[0] == NUM:
            value = int(self.cur[1])
            self.scan_next()
        elif self.cur[0] == LPAR:
            self.scan_next()
            value = self.expression()
            if self.cur[0] != RPAR:
                print("Error: found {} expected {}".format(self.cur[0], RPAR))
                exit(1)
            self.scan_next()
        else:
            print("Error: found type {}, expected {} or {}".format(self.cur[0], NUM, LPAR))
            exit(1)

        return value

    # expression = factor {"+" factor | "*" factor}*
    def expression(self):
        value = self.factor()
        while True:
            if self.cur[0] == ADD:
                self.scan_next()
                value += self.factor()
            elif self.cur[0] == MUL:
                self.scan_next()
                value *= self.factor()
            else:
                break
        return value

# expression = term {"+" term}*
# term = factor {"*" factor}*
# factor = number | "(" expression ")"
class Grammar2(Grammar1):
    # term = factor {"*" factor}*
    def term(self):
        value = self.factor()
        while self.cur[0] == ADD:
            self.scan_next()
            value += self.factor()
        return value

    # expression = term {"+" term}*
    def expression(self):
        value = self.term()
        while self.cur[0] == MUL:
            self.scan_next()
            value *= self.term()
        return value

g1 = Grammar1()
g2 = Grammar2()

assert g1.eval("1 + 2 * 3 + 4 * 5 + 6") == 71
assert g1.eval("5 + (8 * 3 + 9 + 3 * 4 * 3)") == 437
assert g1.eval("5 * 9 * (7 * 3 * 3 + 9 * 3 + (8 + 6 * 4))") == 12240
assert g1.eval("((2 + 4 * 9) * (6 + 9 * 8 + 6) + 6) + 2 + 4 * 2") == 13632
assert g2.eval("1 + 2 * 3 + 4 * 5 + 6") == 231
assert g2.eval("1 + (2 * 3) + (4 * (5 + 6))") == 51
assert g2.eval("2 * 3 + (4 * 5)") == 46
assert g2.eval("5 * 9 * (7 * 3 * 3 + 9 * 3 + (8 + 6 * 4))") == 669060
assert g2.eval("((2 + 4 * 9) * (6 + 9 * 8 + 6) + 6) + 2 + 4 * 2") == 23340

part1, part2 = 0, 0
with open("input", "rt") as f:
    for line in f:
        part1 += g1.eval(line)
        part2 += g2.eval(line)

print("Part1:", part1)
print("Part2:", part2)

File: day18/day18.cpp
#include <cctype>
#include <fstream>
#include <fmt/format.h>

using namespace std;

enum class TokenType
{
	NUM,
	ADD,
	MUL,
	LPAR,
	RPAR,
	END,
};

struct Token
{
	TokenType type;
	string value;
};

// expression = factor {"+" factor | "*" factor}*
// factor = number | "(" expression ")"
class Grammar1
{
public:
	Grammar1() : pos(0), cur{TokenType::END, ""} {}

	intmax_t eval(const string& txt)
	{
		pos = 0;
		scan_next(txt);
		return expression(txt);
	}

protected:
	void scan_next(const string& txt)
	{
		while (pos < txt.size() && isspace(txt[pos]))
		{
			pos++;
		}
		if (pos == txt.size())
		{
			cur.type = TokenType::END;
		}
		else if (isdigit(txt[pos]))
		{
			size_t start = pos;
			while (isdigit(txt[pos]))
			{
				pos++;
			}
			cur.type = TokenType::NUM;
			cur.value = txt.substr(start, pos-start);
		}
		else if (txt[pos] == '+')
		{
			cur.type = TokenType::ADD;
			pos++;
		}
		else if (txt[pos] == '*')
		{
			cur.type = TokenType::MUL;
			pos++;
		}
		else if (txt[pos] == '(')
		{
			cur.type = TokenType::LPAR;
			pos++;
		}
		else if (txt[pos] == ')')
		{
			cur.type = TokenType::RPAR;
			pos++;
		}
		else
		{
			fmt::print(stderr, "Unknown character: {}\n", txt[pos]);
			exit(1);
		}
	}

	// factor = number | "(" expression ")"
	intmax_t factor(const string& txt)
	{
		intmax_t v;
		if (cur.type == TokenType::NUM)
		{
			v = stoll(cur.value);
			scan_next(txt);
		}
		else if (cur.type == TokenType::LPAR)
		{
			scan_next(txt);
			v = expression(txt);
			if (cur.type != TokenType::RPAR)
			{
				fmt::print(stderr, "Error: found {}, expected {}\n",
					   static_cast<int>(cur.type),
					   static_cast<int>(TokenType::RPAR));
				exit(1);
			}
			scan_next(txt);
		}
		else
		{
			fmt::print(stderr, "Error: found {}, expected {} or {}\n",
				   static_cast<int>(cur.type),
				   static_cast<int>(TokenType::RPAR),
				   static_cast<int>(TokenType::LPAR));
			exit(1);
		}
		return v;
	}

	// expression = factor {"+" factor | "*" factor}*
	virtual intmax_t expression(const string& txt)
	{
		intmax_t v = factor(txt);
		while (true)
		{
			if (cur.type == TokenType::ADD)
			{
				scan_next(txt);
				v += factor(txt);
			}
			else if (cur.type == TokenType::MUL)
			{
				scan_next(txt);
				v *= factor(txt);
			}
			else
			{
				break;
			}
		}
		return v;
	}

	size_t pos;
	Token cur;
};

// expression = term {"*" term}*
// term = factor {"+" factor}*
// factor = number | "(" expression ")"
class Grammar2: public Grammar1
{
protected:
	// term = factor {"+" factor}*
	intmax_t term(const string& txt)
	{
		intmax_t v = factor(txt);
		while (cur.type == TokenType::ADD)
		{
			scan_next(txt);
			v += factor(txt);
		}
		return v;
	}

	// expression = term {"*" term}*
	intmax_t expression(const string& txt) override
	{
		intmax_t v = term(txt);
		while (cur.type == TokenType::MUL)
		{
			scan_next(txt);
			v *= term(txt);
		}
		return v;
	}
};

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	intmax_t part1 = 0;
	intmax_t part2 = 0;
	Grammar1 g1;
	Grammar2 g2;
	string line;
	while (getline(input, line))
	{
		part1 += g1.eval(line);
		part2 += g2.eval(line);
	}
	input.close();

	fmt::print("Part1: {}\nPart2: {}\n", part1, part2);
	return 0;
}
